### PR TITLE
feat(orc8r): increase workgroups size if blue-green-orc8r is enabled

### PR DESCRIPTION
**Pull request by joary**
Signed-off-by: Joary Paulo Wanzeler Fortuna <joarypl@fb.com>

## Summary

In terraform the [eks.worker_group](https://github.com/terraform-aws-modules/terraform-aws-eks#input_worker_groups) defines the size and amount of EC2 machines to be deployed in a given k8s cluster, orc8r by default uses 3 x t3.larget.
When deploying a blue-green orc8r setup a different set of machines is needed with more machines of smaller size 5 x t3.medium.

In order to maintain backwards compatibility it is proposed to have a flag indicating the a blue-green deployment is needed. Activation of this flag forces the usage of a different terraform variable pre-configured for blue-green deployment.

1. Include two new variables in orc8r-aws terraform configurations:
    1. Flag to enabled the blue-green deployment
    2. WorkerGroup configuration defaults for a blue-green deployment
1. Configure the EKS module to use blue-green WG values if `enable_orc8r_blue_green_deployment`is active

## Test Plan

With main.tf:

```
module "orc8r" {
   ...
   enable_orc8r_blue_green_deployment = true 
   ...
}
```

terraform plan shows:
```
  # module.orc8r.module.eks.aws_autoscaling_group.workers[0] will be created
  + resource "aws_autoscaling_group" "workers" {
      + arn                       = (known after apply)
      + availability_zones        = (known after apply)
      + default_cooldown          = (known after apply)
      + desired_capacity          = 5
      + force_delete              = false
      + force_delete_warm_pool    = false
      + health_check_grace_period = 300
      + health_check_type         = (known after apply)
      + id                        = (known after apply)
      + launch_configuration      = (known after apply)
      + max_instance_lifetime     = 0
      + max_size                  = 5
      + metrics_granularity       = "1Minute"
      + min_size                  = 1
      + name                      = (known after apply)
      + name_prefix               = "orc8r-wg-1"
      + protect_from_scale_in     = false
      + service_linked_role_arn   = (known after apply)
      + suspended_processes       = [
          + "AZRebalance",
        ]
      + termination_policies      = []
      + vpc_zone_identifier       = [
          + "subnet-0a59aa8e9d94571cd",
          + "subnet-0b488fb0d17e3b66f",
          + "subnet-0ea68e1c0df662ec2",
        ]
      + wait_for_capacity_timeout = "10m"

      + tag {
          + key                 = "Name"
          + propagate_at_launch = true
          + value               = "orc8r-wg-1-eks_asg"
        }
      + tag {
          + key                 = "k8s.io/cluster/orc8r"
          + propagate_at_launch = true
          + value               = "owned"
        }
      + tag {
          + key                 = "kubernetes.io/cluster/orc8r"
          + propagate_at_launch = true
          + value               = "owned"
        }
    }
```

With main.tf:

```
module "orc8r" {
   ...
   enable_orc8r_blue_green_deployment = false # Or not existing 
   ...
}
```

```
  # module.orc8r.module.eks.aws_autoscaling_group.workers[0] will be created
  + resource "aws_autoscaling_group" "workers" {
      + arn                       = (known after apply)
      + availability_zones        = (known after apply)
      + default_cooldown          = (known after apply)
      + desired_capacity          = 3
      + force_delete              = false
      + force_delete_warm_pool    = false
      + health_check_grace_period = 300
      + health_check_type         = (known after apply)
      + id                        = (known after apply)
      + launch_configuration      = (known after apply)
      + max_instance_lifetime     = 0
      + max_size                  = 3
      + metrics_granularity       = "1Minute"
      + min_size                  = 1
      + name                      = (known after apply)
      + name_prefix               = "orc8r-wg-1"
      + protect_from_scale_in     = false
      + service_linked_role_arn   = (known after apply)
      + suspended_processes       = [
          + "AZRebalance",
        ]
      + termination_policies      = []
      + vpc_zone_identifier       = [
          + "subnet-0a59aa8e9d94571cd",
          + "subnet-0b488fb0d17e3b66f",
          + "subnet-0ea68e1c0df662ec2",
        ]
      + wait_for_capacity_timeout = "10m"

      + tag {
          + key                 = "Name"
          + propagate_at_launch = true
          + value               = "orc8r-wg-1-eks_asg"
        }
      + tag {
          + key                 = "k8s.io/cluster/orc8r"
          + propagate_at_launch = true
          + value               = "owned"
        }
      + tag {
          + key                 = "kubernetes.io/cluster/orc8r"
          + propagate_at_launch = true
          + value               = "owned"
        }
    }
```

AWS EC2 instances after deployment with `enable_orc8r_blue_green_deployment` enabled (5x t3.medium)

image

## Additional Information

- [ ] This change is backwards-breaking

### CODE REVIEW

File: orc8r/cloud/deploy/terraform/orc8r-aws/eks.tf [hunk 1]
```diff
@@ -49,7 +49,11 @@ module "eks" {
   }
   worker_additional_security_group_ids = concat([aws_security_group.default.id], var.eks_worker_additional_sg_ids)
   workers_additional_policies          = var.eks_worker_additional_policy_arns
-  worker_groups                        = var.thanos_enabled ? concat(var.eks_worker_groups, var.thanos_worker_groups) : var.eks_worker_groups
+  worker_groups                        = (
```
Comment (hcgatewood) [inline orc8r/cloud/deploy/terraform/orc8r-aws/eks.tf]: What's this line doing? Is a bit hard to follow

**Review by hcgatewood — COMMENTED** [review]
As someone who isn't familiar with worker groups etc: can you flesh out the PR summary a bit to educate me (or e.g. with a pointer), then re-request my review and I'll check it out 🙂

File: orc8r/cloud/deploy/terraform/orc8r-aws/eks.tf (hunk 1 above)
Comment (joary) [reply orc8r/cloud/deploy/terraform/orc8r-aws/eks.tf]: The idea is to substitute the existing `var.eks_worker_groups` by a conditional check based on the `enable_orc8r_blue_green_deployment` flag.

Because there's already a conditional check I had to nest another one. Though thinking now, I can make it more readable using a local variable, something like:

```
locals {
  orc8r_worker_group = var.enable_orc8r_blue_green_deployment ? var.blue_green_worker_groups : var.eks_worker_groups
}

...

worker_groups = var.thanos_enabled ? concat(local.orc8r_worker_group, var.thanos_worker_groups) : local.orc8r_worker_group
```

**Review by hcgatewood — APPROVED** [review]
Awesome! Just handle the misspell GH annotation, otherwise lgtm

File: orc8r/cloud/deploy/terraform/orc8r-aws/variables.tf [hunk 2]
```diff
@@ -134,6 +141,26 @@ variable "thanos_worker_groups" {
...
+  description = "Worker group configuration for EKS. Default value is 1 worker group consisting of 5 t3.medium instances."
+  type        = any
+  default = [
+    {
+      name                 = "wg-1"
+      instance_type        = "t3.medium"
+      asg_desired_capacity = 5
+      asg_min_size         = 1
+      asg_max_size         = 5
+      autoscaling_enabled  = false
```
Comment (arunuke) [inline orc8r/cloud/deploy/terraform/orc8r-aws/variables.tf:158]: Should we consider enabling autoscaling_enabled to 'true' with desired capacity reduced to 3 or 4 (as the number of supported pods per node)? This will enable us to have a desired capacity that is lesser than max, but have the ability to automatically scale when the 'Green' cluster is deployed.

**Review by arunuke — CHANGES_REQUESTED** [review]

**Review by arunuke — APPROVED** [review]
LGTM !!

File: orc8r/cloud/deploy/terraform/orc8r-aws/variables.tf (hunk 2 above)
Comment (arunuke) [reply orc8r/cloud/deploy/terraform/orc8r-aws/variables.tf:158]: We will start with the default size as we discussed while enabling autoscaling in a follow up PR.

**Comment by joary** [comment]
@arunuke @hcgatewood , I rebased this PR, though during validation I found that 5 x t3.medium will not be enough to deploy two full Orc8r.

The full blue-green deployment consumes 129 pods [see the full list](https://gist.github.com/joary/f954522a4e045d4ae7973df8c4432a17)

From AWS docs a [t3.medium can handle 17 pods](https://github.com/awslabs/amazon-eks-ami/blob/master/files/eni-max-pods.txt#L406). Then we would need 8 x t3.medium. 

I'm updating the PR with that value since this will be a blocker anyway.

**Comment by joary** [comment]
Also had to add new variables to `orc8r/cloud/deploy/orc8r_deployer/docker/root/data/vars.yml` in order to clear required test `cloud-workflow / cloud-tests`

**Review by arunuke — APPROVED** [review]
LGTM.
